fix: --- kubescape fix

XDRAGON2002 opened the pull request:
## Overview

As mentioned in the issue, in case of a `---` being present in the beginning of the yaml manifest file the `kubescape fix` command doesn't seem to work properly and throws an error. This behavior has been fixed now.

## Additional Information

The problem is related to the `yaml/v3` dependency we use for parsing/writing yaml which it does by converting the yaml indents into corresponding `yaml.Node` objects of one of 5 possible types, but marshalling is only supported for 3 of those kinds, due to `---` being present at the beginning the dependency assigns it a node kind which cannot be marshalled hence the error.
Best option would be to get Kubescape to ignore that initial `---` while processing. 

## How to Test

1. Create a `resource.yaml` file like:
```yaml
---
apiVersion: v1
kind: Pod
metadata:
  name: pod
spec:
  containers:
    - name: container
      image: image:latest
```
2. Run `kubescape scan resource.yaml --format json --output output.json`
3. Run `kubescape fix output.json`

## Related issues/PRs:

Resolved #1255 

## Checklist before requesting a review

- [x] My code follows the style guidelines of this project
- [x] I have commented on my code, particularly in hard-to-understand areas
- [x] I have performed a self-review of my code
- [x] If it is a core feature, I have added thorough tests.
- [x] New and existing unit tests pass locally with my changes

vladklokun commented on core/pkg/fixhandler/fixhandler.go:
```diff
@@ -208,6 +208,12 @@ func (h *FixHandler) ApplyChanges(ctx context.Context, resourcesToFix []Resource
 			continue
 		}
 
+		// Handle the case where the resource file starts with ---
+		// causes yaml.Node to misinterpret the resources
+		if fileAsString[:3] == "---" {
+			fileAsString = "# " + fileAsString
+		}
+
```
issue: YAML file sanitization is not abstracted away

A function should not deal with implementation details of more than one logical thing at once. This allows it to be testable, maintainable and reduce the amount of things a reader has to know when looking at a function. That is, improve local reasoning.

However, the suggested code tries to do something seemingly small and clever on a low level of abstraction in a function that deals with a high level of abstraction. In other words, before this change, the `ApplyChanges` method knew how to fix resources using a set of lower-level functions that handled the details. It did one thing and, ideally, did it well.

After this change, this method would have to know how to fix resources AND what are the exact steps to sanitize a YAML file so `yaml/v3` will not complain. But there’s no need for this.

We would benefit from having a separate function / method / struct that sanitizes an input file. That way, we can test this separate function and be confident that it fulfills the contract it provides. So `ApplyChanges` can utilize, say, `SanitizeYAML` without knowing how to actually sanitize it.

XDRAGON2002 commented:
@vladklokun Have updated the PR with the requested changes.
Moved the sanitization functionalities to separate functions of their own, added test for a yaml file with `---` as the beginning, also added tests for the abstracted sanitization functions.

Also found a typo? in the tests regarding the `removals` so fixed that too.

Let me know if there are any more changes required! Thanks! ;)

kiratsingh98 commented on core/pkg/fixhandler/fixhandler.go:
```diff
@@ -368,3 +372,20 @@ func determineNewlineSeparator(contents string) string {
+
+// For the --- case to ensure correct output file format
+func revertSanitizeYaml(fixedYamlString string) string {
+	if fixedYamlString[:5] == "# ---" {
```
@XDRAGON2002 In case there are multiple resources (say, a Service, a Deployment, and a Pod), does this function handle that aspect? It should be generic enough to test on different [yamls.](https://raw.githubusercontent.com/kubernetes/website/main/content/en/examples/application/nginx-app.yaml)

// @vladklokun @craigbox @slashben

XDRAGON2002 replied:
@kiratsingh98 Yes it does, because the problem only occurs when the first line in the yaml is `---` so if the separator is used anywhere apart from that it's the same exact expected behavior as before.

As evident from the code changes as well, we're only concerned when the manifest begins with `---` everything else works fine.

vladklokun commented on core/pkg/fixhandler/fixhandler_test.go:
```diff
@@ -162,27 +169,41 @@ func getTestCases() []indentationTestCase {
-			getTestDataPath := func(filename string) string {
-				currentFile := "testdata/" + filename
-				return filepath.Join(testutils.CurrentDir(), currentFile)
+			inputFilename := getTestDataPath(tc.inputFile)
+			input, err := os.ReadFile(inputFilename)
+			if err != nil {
+				t.Fatalf(`Unable to open file %s due to: %v`, inputFilename, err)
+			}
+			expectedFilename := getTestDataPath(tc.expectedFile)
+			wantRaw, err := os.ReadFile(expectedFilename)
+			if err != nil {
+				t.Fatalf(`Unable to open file %s due to: %v`, expectedFilename, err)
```
praise: great catch with the `testdata/removes` typo!

This led me to check why didn’t the tests fail if the file did not exist, and I noticed that we silently ignored the error. Fixed that.

XDRAGON2002 commented:
Since we already have test files in the `hybrid` test data folder now, I guess we can go ahead and remove the files I had added to the `insert` test data folder?

vladklokun commented on core/pkg/fixhandler/testdata/inserts/tc-12-00-begin-with-document-separator.yaml:
```diff
@@ -0,0 +1,10 @@
+---
+apiVersion: v1
+kind: Pod
+metadata:
+  name: begin-with-document-separator
+
+spec:
+  containers:
+  - name: nginx_container
+    image: nginx
```
praise: great decision to add a new test case file!

This decision is good in particular because:
- it follows the existing test conventions for this functionality
- the existing conventions test our public-facing behavior and contract: that we take in a specific file, apply some fixes and return some specific expected file.

So this new test file will do the best thing tests are supposed to do: check that we keep the promises we make to the public.

Now, you might notice I made some changes to the PR. Among these changes is that I moved the calls to the sanitizer functions to the `ApplyFixToContent` function. This is because our aforementioned public-facing behavior is provided by the `ApplyFixToContent` function.

So if the `ApplyFixToContent` function promises the following contract:

> given a valid YAML file and a YAML expression, I will apply the provided fixes to it.

it only makes sense that this function would include the sanitization behavior in itself. Now, we just need to make sure that we test the case of YAML files that start with the document separator. I removed the tests of the unexported (private) sanitizer functions for this reason as well. There is no reason to test the brittle details of the implementation as long as we test the public contracts.

So the takeaway here is that the best practice is to test your contracts aka behavior that you promise to uphold to the clients.

vladklokun replied:
@XDRAGON2002 
> Since we already have test files in the `hybrid` test data folder now, I guess we can go ahead and remove the files I had added to the `insert` test data folder?

The cases are different, so let’s keep the one you’ve added. The more meaningful cases, the better.

vladklokun reviewed (approved):
Great job, LGTM!

I apologize for making changes to your PR myself. We wanted to merge this promptly, so I wanted to skip the back-and-forth. That’s how important of a thing you have fixed :)

You did well, thank you for your contribution @XDRAGON2002!

XDRAGON2002 commented:
> Great job, LGTM!
> 
> I apologize for making changes to your PR myself. We wanted to merge this promptly, so I wanted to skip the back-and-forth. That’s how important of a thing you have fixed :)
> 
> You did well, thank you for your contribution @XDRAGON2002!

Not a problem at all @vladklokun!
I was just going through the commits as you were making them, this to be honest is one of the best things about opensource, as a lesser experienced developer it allows me to directly glimpse into how a more experienced developer might think and make implementation decisions.
This was a great exercise for the same! Thanks a lot!

vladklokun commented:
> As for the function / struct that sanitizes the input YAML, you should have a self-contained unit test

@XDRAGON2002 lol just realized I might have given you mixed info on whether to have the unit test for the sanitizer function or not.

To clarify my point, it’s okay to write the unit test to help you develop a private function, class or whatever. Personally, I prefer to do it this way. But these tests should be removed before merging the changes since they make it more difficult to refactor code in the future. At least, that’s what the books say :)

There’s a great [talk](https://www.youtube.com/watch?v=EZ05e7EMOLM) on Test-driven Development that covers this point of testing private and public things, I recommend watching it.

alegrey91 reviewed (approved):
LGTM.
Great job @XDRAGON2002 !

XDRAGON2002 commented:
> > As for the function / struct that sanitizes the input YAML, you should have a self-contained unit test
> 
> @XDRAGON2002 lol just realized I might have given you mixed info on whether to have the unit test for the sanitizer function or not.
> 
> To clarify my point, it’s okay to write the unit test to help you develop a private function, class or whatever. Personally, I prefer to do it this way. But these tests should be removed before merging the changes since they make it more difficult to refactor code in the future. At least, that’s what the books say :)
> 
> There’s a great [talk](https://www.youtube.com/watch?v=EZ05e7EMOLM) on Test-driven Development that covers this point of testing private and public things, I recommend watching it.

Ahh, makes a lot more sense know, because I was just puzzled regarding the fact that why were the tests removed since I was asked to add them XD
